File: web/dashboard/src/lib/escrow.ts
import type { AuthSession } from "./auth";
import { serviceRequest } from "./serviceApi";

export interface EscrowGuardian {
  id: string;
  name: string;
  email: string;
  organization: string;
  notary_cert_fingerprint?: string;
  status: "active" | "suspended";
  added_at: string;
}

export interface EscrowPolicy {
  id: string;
  tenant_id: string;
  name: string;
  description: string;
  key_filter: string;
  threshold: number;
  guardian_ids: string[];
  legal_hold: boolean;
  jurisdiction: string;
  enabled: boolean;
  created_at: string;
  escrow_count: number;
}

export interface EscrowedKey {
  id: string;
  policy_id: string;
  policy_name: string;
  key_id: string;
  key_name: string;
  algorithm: string;
  escrowed_at: string;
  escrowed_by: string;
  guardian_ids: string[];
  status: "active" | "under_recovery" | "recovered" | "destroyed";
}

export interface RecoveryRequest {
  id: string;
  escrow_id: string;
  key_id: string;
  key_name: string;
  requestor: string;
  reason: string;
  legal_reference?: string;
  status: "pending" | "approved" | "denied" | "completed";
  approvals: { guardian_id: string; guardian_name: string; approved_at?: string; denied_at?: string }[];
  required_approvals: number;
  created_at: string;
  completed_at?: string;
}

export async function listGuardians(session: AuthSession): Promise<EscrowGuardian[]> {
  const res = await serviceRequest<any>(session, "keycore", "/escrow/guardians");
  return res.data ?? [];
}

export async function addGuardian(session: AuthSession, data: Partial<EscrowGuardian>): Promise<EscrowGuardian> {
  const res = await serviceRequest<any>(session, "keycore", "/escrow/guardians", { method: "POST", body: JSON.stringify(data) });
  return res.data ?? res;
}

export async function listPolicies(session: AuthSession): Promise<EscrowPolicy[]> {
  const res = await serviceRequest<any>(session, "keycore", "/escrow/policies");
  return res.data ?? [];
}

export async function createPolicy(session: AuthSession, data: Partial<EscrowPolicy>): Promise<EscrowPolicy> {
  const res = await serviceRequest<any>(session, "keycore", "/escrow/policies", { method: "POST", body: JSON.stringify(data) });
  return res.data ?? res;
}

export async function listEscrowedKeys(session: AuthSession): Promise<EscrowedKey[]> {
  const res = await serviceRequest<any>(session, "keycore", "/escrow/keys");
  return res.data ?? [];
}

export async function escrowKey(session: AuthSession, data: {
  policy_id: string;
  policy_name: string;
  key_id: string;
  key_name: string;
  algorithm: string;
  guardian_ids: string[];
  escrowed_by?: string;
}): Promise<EscrowedKey> {
  const res = await serviceRequest<any>(session, "keycore", "/escrow/keys", { method: "POST", body: JSON.stringify(data) });
  return res.data ?? res;
}

export async function listRecoveryRequests(session: AuthSession): Promise<RecoveryRequest[]> {
  const res = await serviceRequest<any>(session, "keycore", "/escrow/recovery");
  return res.data ?? [];
}

export async function createRecoveryRequest(session: AuthSession, data: { escrow_id: string; reason: string; legal_reference?: string }): Promise<RecoveryRequest> {
  const res = await serviceRequest<any>(session, "keycore", "/escrow/recovery", { method: "POST", body: JSON.stringify(data) });
  return res.data ?? res;
}

export async function approveRecovery(session: AuthSession, requestId: string): Promise<RecoveryRequest> {
  const res = await serviceRequest<any>(session, "keycore", `/escrow/recovery/${requestId}/approve`, { method: "POST" });
  return res.data ?? res;
}

export async function denyRecovery(session: AuthSession, requestId: string, reason: string): Promise<RecoveryRequest> {
  const res = await serviceRequest<any>(session, "keycore", `/escrow/recovery/${requestId}/deny`, { method: "POST", body: JSON.stringify({ reason }) });
  return res.data ?? res;
}

File: services/keycore/escrow.go
package main

import "time"

// EscrowGuardian is a trusted party who holds a share of an escrowed key.
type EscrowGuardian struct {
	ID                   string    `json:"id"`
	TenantID             string    `json:"tenant_id"`
	Name                 string    `json:"name"`
	Email                string    `json:"email"`
	Organization         string    `json:"organization"`
	NotaryCertFingerprint string   `json:"notary_cert_fingerprint"`
	Status               string    `json:"status"`
	AddedAt              time.Time `json:"added_at"`
}

// EscrowPolicy defines which keys are escrowed and how many guardians must
// approve a recovery request.
type EscrowPolicy struct {
	ID             string    `json:"id"`
	TenantID       string    `json:"tenant_id"`
	Name           string    `json:"name"`
	Description    string    `json:"description"`
	KeyFilter      string    `json:"key_filter"`
	Threshold      int       `json:"threshold"`
	GuardianIDs    []string  `json:"guardian_ids"`
	LegalHold      bool      `json:"legal_hold"`
	Jurisdiction   string    `json:"jurisdiction"`
	Enabled        bool      `json:"enabled"`
	CreatedAt      time.Time `json:"created_at"`
	EscrowCount    int       `json:"escrow_count"`
}

// EscrowedKey represents a key that has been placed into escrow under a policy.
type EscrowedKey struct {
	ID          string    `json:"id"`
	TenantID    string    `json:"tenant_id"`
	PolicyID    string    `json:"policy_id"`
	PolicyName  string    `json:"policy_name"`
	KeyID       string    `json:"key_id"`
	KeyName     string    `json:"key_name"`
	Algorithm   string    `json:"algorithm"`
	GuardianIDs []string  `json:"guardian_ids"`
	Status      string    `json:"status"`
	EscrowedAt  time.Time `json:"escrowed_at"`
	EscrowedBy  string    `json:"escrowed_by"`
}

// RecoveryApproval records a single guardian's approval or denial of a
// recovery request.
type RecoveryApproval struct {
	GuardianID string    `json:"guardian_id"`
	Decision   string    `json:"decision"`
	DecidedAt  time.Time `json:"decided_at"`
	Notes      string    `json:"notes,omitempty"`
}

// RecoveryRequest is a request to recover an escrowed key.
type RecoveryRequest struct {
	ID               string             `json:"id"`
	TenantID         string             `json:"tenant_id"`
	EscrowID         string             `json:"escrow_id"`
	KeyID            string             `json:"key_id"`
	KeyName          string             `json:"key_name"`
	Requestor        string             `json:"requestor"`
	Reason           string             `json:"reason"`
	LegalReference   string             `json:"legal_reference"`
	Status           string             `json:"status"`
	RequiredApprovals int               `json:"required_approvals"`
	Approvals        []RecoveryApproval `json:"approvals"`
	CreatedAt        time.Time          `json:"created_at"`
	CompletedAt      *time.Time         `json:"completed_at,omitempty"`
}
